src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -287,11 +293,32 @@ private void TlsHandshakeStopCore(in ConnectionMetricsContext metricsContext, lo
 
     private static void InitializeConnectionTags(ref TagList tags, in ConnectionMetricsContext metricsContext)
     {
-        if (metricsContext.ConnectionContext.LocalEndPoint is { } localEndpoint)
+        var localEndpoint = metricsContext.ConnectionContext.LocalEndPoint;
+        if (localEndpoint is IPEndPoint localIPEndPoint)
+        {
+            tags.Add("server.socket.address", localIPEndPoint.Address.ToString());
+            tags.Add("server.socket.port", localIPEndPoint.Port);
+
+            // There isn't an easy way to detect whether QUIC is the underlying transport.
+            // This code assumes that a multiplexed connection is QUIC.
+            // Improve in the future if there are additional multiplexed connection types.
+            var transport = metricsContext.ConnectionContext is not MultiplexedConnectionContext ? "tcp" : "udp";
+            tags.Add("network.transport", transport);
+        }
+        else if (localEndpoint is UnixDomainSocketEndPoint udsEndPoint)
+        {
+            tags.Add("server.socket.address", udsEndPoint.ToString());
+            tags.Add("network.transport", "unix");
+        }
+        else if (localEndpoint is NamedPipeEndPoint namedPipeEndPoint)
+        {
+            tags.Add("server.socket.address", namedPipeEndPoint.ToString());
+            tags.Add("network.transport", "pipe");
+        }
+        else if (localEndpoint != null)
         {
-            // TODO: Improve getting string allocation for endpoint. Currently allocates.
-            // Considering adding a way to cache on ConnectionContext.
-            tags.Add("endpoint", localEndpoint.ToString());
+            tags.Add("server.socket.address", localEndpoint.ToString());
+            tags.Add("network.transport", localEndpoint.GetType().Name);
```

Comment:
maybe `localEndpoint.[AddressFamily](https://learn.microsoft.com/en-us/dotnet/api/system.net.endpoint.addressfamily?view=net-7.0#system-net-endpoint-addressfamily)` as a shorter and cross-language alternative?